README.md
<img src="https://user-images.githubusercontent.com/25137010/46918261-65087380-cfd0-11e8-8a47-5810dc217eb2.png" width="15%"></img> <img src="https://user-images.githubusercontent.com/25137010/46918262-65087380-cfd0-11e8-99fe-135033d2a976.png" width="15%"></img> <img src="https://user-images.githubusercontent.com/25137010/46918263-65087380-cfd0-11e8-9127-bda7cbaea364.png" width="15%"></img> 

# rartracker
A modern front and backend bittorrent tracker written in AngularJS and PHP.
## Current version
0.3.5
## Author
the swedish torrent king / ERR1R
## Feature highlights
* Super responsive GUI thanks to the SPA-application nature of AngularJS
* Advanced admin-features for monitoring site activity, handling reports and multi-deleting torrents.
* Many features suitable for scene only trackers
* Awesome features like bonus-system, leech bonus, seed-time, request system and rss-system.
* Highly skinnable and mobile friendly design with Bootstrap CSS
* Very clean code to grasp for developers
* Dynamic multi language support

# Installation
## 1. Install Node.js/npm and Git

**Windows:**
* https://nodejs.org/en/download/
* https://git-scm.com/download/win (Choose: Use/run Git from the Command Prompt (cmd.exe))

**Ubuntu/Debian:**
```sh
$ sudo apt-get install -y nodejs
$ sudo apt-get install -y git-core
```

## 2. Install dependencies
From the project folder install all build script packages and all 3rd party dependencies
```sh
$ npm install --global bower
$ npm install --global gulp-cli
$ npm install
$ bower install --allow-root <-- keep the allow-root flag, if logged in as root
$ npm run dist
```
## 3. Import database (databse/database.sql)

# Configurations

### 1. Basic config
Create the file **secrets.php** in the api/ folder.
````
<?php
$database = 'mysql';
$host = '127.0.0.1';
$dbname = 'rartracker';
$username = 'root';
$password = '';
````

Some site settings in api/Config.php and app/app.config.js should be changed!

### 2. Generate new unique salt hashes (optional for security)
* Note that updating the salts will make current registered accounts unusable, you'd want to log in and create invite codes (step 4) before changing salts and finally registering new accounts.
* $passwordSalt and $cookieSalt in **User.php** should be updated with new random hashes.

### 3. Create admin accounts
Use built in account named "**System**" password: "**password**" to create invites and register new admin-accounts
The "System" account **must remain** as a parked account and have Staff rights because it's used as deleter of torrents, sender of PM and creation of forum topics etc. Change System password.

## Server settings
### Recommended packages
Apache:
``php7.2 apache2 mariadb-server libapache2-mod-php7.2 php7.2-mysql memcached php7.2-curl php7.2-memcached php-msgpack php-igbinary``
## Only for apache2
### Enable rewrite module and change AllowOverride from "None" to "All" in httpd.conf in order for .htaccess to work
``a2enmod rewrite``

Nginx:
``php7.2 nginx mariadb-server libapache2-mod-php7.2 php7.2-mysql memcached php7.2-curl php7.2-memcached php-msgpack php-igbinary php7.2-fpm``

Tested with mariadb-server-10.3 / php7.2/php7.2-fpm / apache/ nginx


### Permissions
The following folders needs to be created and be given write permission:

* torrents/
* img/imdb/

### Recommended MariaDB settings
This is for making the fulltext search work. Use innoDB database and add below lines in your my.cnf under [mysqld].
Also disable the stopword function with innodb_ft_enable_stopword = OFF.
```sh
innodb_ft_enable_stopword = OFF
innodb_ft_min_token_size=1
ft_min_word_len=1

```

## Nginx configuration
```sh
server {
#simple config
	listen 80;
	listen 443 ssl;
	server_name digitalcore.nl;
   
       ssl_certificate     /etc/nginx/ssl/digitalcore.nl.crt;
	ssl_certificate_key /etc/nginx/ssl/digitalcore.nl.key;

	root /home/www/rartor/rartracker;
	index index.html;

	location / { try_files $uri /index.html;}
	location ~ ./img { } 
	location ~ ./phpMyAdmin { } 
	location ~ .(html)$ { } 
	rewrite ^/api/v1/(.*)$ /api/api-v1.php?url=$1 last;

	location ~ \.php$ {
       include snippets/fastcgi-php.conf;
       fastcgi_pass unix:/run/php/php7.2-fpm.sock;
       }
}

```

## Apache Configuration
```sh 
<VirtualHost domainname.nl:443>
    ServerName domainname.nl
    DocumentRoot /var/www/rartracker
    LogLevel warn
    ServerSignature Off
    ErrorLog /var/log/apache2/error.log
    <Directory "/var/www/rartracker">
        AllowOverride All
        Require all granted
    RewriteEngine On
    </Directory>
SSLEngine on
SSLCertificateFile /etc/apache2/ssl/certname_nl.crt
SSLCertificateKeyFile /etc/apache2/ssl/certname.key
SSLCertificateChainFile /etc/apache2/ssl/certname_nl.ca-bundle 
</VirtualHost>

<VirtualHost domainname.nl:80>
    ServerName domainname.nl
    DocumentRoot /var/www/rartracker
    LogLevel warn
    ServerSignature Off
    ErrorLog /var/log/apache2/error.log
    <Directory "/var/www/rartracker">
        AllowOverride All
        Require all granted
    RewriteEngine On
    </Directory>
</VirtualHost>
```
## General Configurations

You'll need to listen on:

Port 80 to reach the site
Port 1337 so torrent clients can talk to the tracker
Port 1338 ( with SSL enabled! ) so torrent clients can securely talk to the tracker
By default this option is enabled for every member
In /api/Config.php you'll need to edit TRACKER_URL and TRACKER_URL_SSL.

```php
<?php
    const TRACKER_URL = "http://domain.nl:1337";
    const TRACKER_URL_SSL = "https://domain.nl:1338"
```

Obviously replace domain.nl with whatever domain you use.

Dev-hint

If you want to force regular HTTP tracker communication, just set TRACKER_URL_SSL to the same value as TRACKER_URL.
This allows you to ignore messing with SSL for a while.

I do recommend setting up SSL if you're using this in production.

## Crontab settings (crontab -e)
```sh
12      *       *       *       *       wget -O /dev/null http://127.0.0.1/api/v1/run-leechbonus
*/20    *       *       *       *       wget -O /dev/null http://127.0.0.1/api/v1/run-cleanup
0       0       *       *       5       wget -O /dev/null http://127.0.0.1/api/v1/run-bonus
0       20      *       *       *       wget -O /dev/null http://127.0.0.1/api/v1/run-statistics
```

## Language support

The default language is English, but if you want the back and front end to default to another of the available languages you need to change the default language variable in **Config.php** and **app.config.js**. If the user changes the display language, the default language will still be used for site wide logs and automatic messages.

## Developing and deploying
### Developing
In the project folder run ``npm run watch``. This will lauch a watcher that continuously build the code into the dist/.
### Deploying
In the project folder run ``npm run dist`` and the code will be minified and scrambled, a JsHint check will also be made.

The following folders should not exist in production. The bundles contains everything needed.
app/
bower_components/
node_modules/

# License
[WTFPL]

[//]: #
[WTFPL]: <http://www.wtfpl.net/>


CHANGELOG.md
# rartracker changelog

# 0.3.5 (2018-10-12)

### New features and improvements

* Removed all swe 
* Removed youtube trailer
* Removed subtitles (who need this)
* Add Jackett indexer file so you can compile jackett for this tracker
* Updated readme
* Added IMDB trailer

### Bugfixes

* Fixed IMDB
* Fixed Watch
* Fixed couchpotato
* Fixed some more stuff

# 0.3.2 (2016-08-27)

### New features and improvements

* Refactored (removed) rss.php, brss.php and subdownload.php into new /api/v1 URLs
* Categories are now specified in Config.php on server side
* API for "torrents matcher" which is a client application for matching and loading torrents into bittorrent clients. (Application soon to be released on this github account)

### Bugfixes

* SweTV fetch now works again
* Genre selector in custom start page edit was missing
* Bootstrap fonts on /start were missing
* Additional language fixes

# 0.3.1 (2016-08-04)

### New features and improvements

* JS-libs and CSS files are now bundled and bower_components are no longer needed in deployment

### Bugfixes

* Language fixed

# 0.3.0 (2016-06-29)

### New features and improvements

* Dynamic multi language support with English as default language, finally!

# 0.2.2 (2016-06-24)

### New features and improvements

* Youtube trailers
* Torrent lists
* Admin can delete IP log entries

### Bugfixes

* Requests lists now save sorted state
* Swesub form
* Requests can now be in new section due to new/archive separation instead of using reqid
* Weekly earned bonus points calculated correctly on profile view
* Wide NFO files are now wrapped
* Swetv data source fetch working again
* Increased topic length
* Related torrents leaked uploader user id

# 0.2.1 (2016-03-10)

### New features and improvements

* Archive torrents section removed. "New" and "archive" torrents are now mixed
* Filters for displaying new/archive, swesub, p2p, freeleech and default settings on user profile
* View filled requests
* Validate invite codes "before" registration.
* Email hashed. (run migrate.php on existing database)
* Watch toplists now based on recent activity

### Bugfixes

* Several disabled buttons were clickable
* Date in "watch" table (bevaka) fixed (run migrate.php), and for displaying more recent toplists data
* Buy -10 GB on friend now works
* Watch torrent page now works

# 0.2.0 (2016-02-26)

### New features and improvements

* Request comments
* Slug urls for requests and forum topics
* Subtitle quality selector
* IMDB info are fetched/guessed before upload
* Improved admin user search
* Multi-delete in Mailbox
* Polls paginated
* Better predb source
* Admins can delete suggestions and updated status will post into forum.
* Improved "IMDB guessing" for TV-series based on release name title

### Bugfixes
* Anyone could delete torrent comments
* Performance improvements
* History/back behaviour when sorting/browsing the site in general now works as expected
* Search info when upl
* Minclassread now works on forumheads
* IP check on signup could match itself
* Snatch log not logged correctly
* "Banlist" on friends-page did not work
* Countless minor bugfixes...

### Architectural and code improvements

* Better Config.php and app.config.js for general configurations
* Switching from grunt to the much faster and easier **gulp** as build script.
* Adding babel(es2015) to build step for es6 support, mostly for fat arrows.
* All JS-files, module and folder stucture has been reorganized and improved alot
* Code splitted into feature modules, now much easier to plug and play modules/features from app.modules.js
* Dropped $scope in favour for controllerAs 'vm'
* Alot of code cleanup in controllers, lots of stuff moved to components instead.
* Taking full use of ui-router URL state params for a proper browser history behaviour
* Upgrading all libs (angular, ui-router, ui-bootstrap etc) and support to latest version.
* Switched from memcache to memcached which actually works
* Fixed php strict errors (~E_STRICT no longer required)

# 0.1.0 (2015-11-01)

* Initial Release


api/Config.php
<?php

class Config {

	const NAME = "DigitalCore.nl";
	const SITE_NAME = "DigitalCore.nl";
	const SITE_URL = "https://digitalcore.nl";
	const SITE_MAIL = "no-reply@digitalcore.nl";

	const SUGGESTION_FORUM_ID = 4;
	const POLLS_FORUM_ID = 3;
	const NEWS_FORUM_ID = 1;

	const DEFAULT_LANGUAGE = "en";
	public static $languages = ["en"];

	const TRACKER_URL = "http://digitalcore.nl:80";
    	const TRACKER_URL_SSL = "https://digitalcore.nl:443";

	public static $userClasses = array(
		0 => "Extra",
		1 => "Actor",
		2 => "Movie star",
		3 => "Director",
		4 => "Producer",
		6 => "Uploader",
		7 => "VIP",
		8 => "Staff");

	public static $categories = array(
		"DVDR_PAL" => array("id" => 1, "name" => "Movies/DVDR"),
              "MOVIE_SD" => array("id" => 2, "name" => "Movies/XviD"),
		"BLURAY" => array("id" => 3, "name" => "Movies/BluRay"),
		"MOVIE_4K" => array("id" => 4, "name" => "Movies/4K"),
		"MOVIE_720P" => array("id" => 5, "name" => "Movies/720p"),
		"MOVIE_1080P" => array("id" => 6, "name" => "Movies/1080p"),
		"TV_720P" => array("id" => 8, "name" => "Tv/720p"),
		"TV_1080P" => array("id" => 9, "name" => "Tv/1080p"),
		"XVID_TV" => array("id" => 10, "name" => "Tv/XVID"),
		"DVDR_TV" => array("id" => 11, "name" => "Tv/DVDR"),
		"MUSIC" => array("id" => 12, "name" => "Music"),
		"UNKNOWN" => array("id" => 17, "name" => "Unknown"),
		"APPS_0DAY" => array("id" => 18, "name" => "Apps/0DAY"),
		"PC_GAMES" => array("id" => 19, "name" => "Games/PC"),
		"PC_APPS" => array("id" => 20, "name" => "Apps/PC")
	);

}
